clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareFetchRequestManagerTest.java

```diff
@@ -0,0 +1,908 @@
+            networkClientDelegate.addAll(pollResult.unsentRequests);
+            return pollResult.unsentRequests.size();
+        }
+    }
+
+    private class TestableNetworkClientDelegate extends NetworkClientDelegate {
+
+        private final Logger log = LoggerFactory.getLogger(NetworkClientDelegate.class);
+        private final ConcurrentLinkedQueue<Node> pendingDisconnects = new ConcurrentLinkedQueue<>();
+
+        public TestableNetworkClientDelegate(Time time,
+                                             ConsumerConfig config,
+                                             LogContext logContext,
+                                             KafkaClient client) {
+            super(time, config, logContext, client);
+        }
+
+        @Override
+        public void poll(final long timeoutMs, final long currentTimeMs) {
+            handlePendingDisconnects();
+            super.poll(timeoutMs, currentTimeMs);
+        }
+
+        public void poll(final Timer timer) {
+            long pollTimeout = Math.min(timer.remainingMs(), requestTimeoutMs);
+            if (client.inFlightRequestCount() == 0)
+                pollTimeout = Math.min(pollTimeout, retryBackoffMs);
+            poll(pollTimeout, timer.currentTimeMs());
+        }
+
+        private Set<Node> unsentRequestNodes() {
+            Set<Node> set = new HashSet<>();
+
+            for (UnsentRequest u : unsentRequests())
+                u.node().ifPresent(set::add);
+
+            return set;
+        }
+
+        private List<UnsentRequest> removeUnsentRequestByNode(Node node) {
+            List<UnsentRequest> list = new ArrayList<>();
+
+            Iterator<UnsentRequest> iter = unsentRequests().iterator();
+
+            while (iter.hasNext()) {
+                UnsentRequest u = iter.next();
+
+                if (node.equals(u.node().orElse(null))) {
+                    iter.remove();
+                    list.add(u);
+                }
+            }
+
+            return list;
+        }
+
+        @Override
+        protected void checkDisconnects(final long currentTimeMs) {
+            // any disconnects affecting requests that have already been transmitted will be handled
+            // by NetworkClient, so we just need to check whether connections for any of the unsent
+            // requests have been disconnected; if they have, then we complete the corresponding future
+            // and set the disconnect flag in the ClientResponse
+            for (Node node : unsentRequestNodes()) {
+                if (client.connectionFailed(node)) {
+                    // Remove entry before invoking request callback to avoid callbacks handling
+                    // coordinator failures traversing the unsent list again.
+                    for (UnsentRequest unsentRequest : removeUnsentRequestByNode(node)) {
+                        // TODO: this should likely emulate what's done in ConsumerNetworkClient
```

Comment:
Hmm. I think this should likely emulate what's done in ConsumerNetworkClient. Let me know if that's not easy to do now.

Reply:
Right, I have added the implementation now.